Add ohrwurm MS management interface

schettn opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->

-   [x] Have you added an explanation of what your changes do and why you'd like them to be included?
-   [x] Have you updated or added documentation for the change?
-   [x] Have you tested your changes with successful results?

**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Documentation (non-breaking change which adds documentation)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the new behavior (if this is a feature change)?**
We have implemented a fully functional management interface for micro services, especially [ohrwurm](https://github.com/snek-shipyard/ohwurm).
The key features of this implementation are:
- Select different microservices
- Project management (add/edit/delete/list))
- Member management (add/edit/delete/list))
- Track management (upload/edit/delete/list)
- Different member roles (supervisor)
- Authentication
- Force password change on first login

#2 #5 #6

The following figures gives a brief impress about the introduces implemetation:

- Service chooser
![image](https://user-images.githubusercontent.com/52858351/103803967-a70aad80-5051-11eb-9546-3214ba3877f3.png)
- Login
![image](https://user-images.githubusercontent.com/52858351/103804025-ba1d7d80-5051-11eb-8fde-45943fb22712.png)
- Password change
![image](https://user-images.githubusercontent.com/52858351/103804264-1b455100-5052-11eb-9d17-facdc332b0f0.png)
- Project table
![image](https://user-images.githubusercontent.com/52858351/103804041-c4d81280-5051-11eb-80cc-4f72d0647f6a.png)
- Project model
![image](https://user-images.githubusercontent.com/52858351/103804119-e76a2b80-5051-11eb-887b-65ec51531cee.png)
- Track table
![image](https://user-images.githubusercontent.com/52858351/103804111-e5a06800-5051-11eb-9fe4-e14c3f061273.png)
- Track model
![image](https://user-images.githubusercontent.com/52858351/103804187-fbae2880-5051-11eb-9bdd-1f3127357da2.png)
- Member table
![image](https://user-images.githubusercontent.com/52858351/103804222-08328100-5052-11eb-9be4-f6c5230280eb.png)
- Member model
![image](https://user-images.githubusercontent.com/52858351/103804239-11235280-5052-11eb-8fb7-a1a1ebbe58ca.png)

schettn committed:
Add general app structure and interface for ohrwurm

This commit introduced the complete viewing interface
for the ohrwurm microservice which is developed under:
https://github.com/snek-shipyard/ohrwurm.

schettn committed:
Apply suggestions from code review

Co-authored-by: Florian Kleber <kleberbaum@erebos.xyz>

schettn committed:
Merge pull request #2 from snek-shipyard/001-add-ohrwurm-interface

Add ohrwurm interface

schettn committed:
Add favicons and remove legacy logo

Several favicons have been added.
The manifest have been adjusted.
A legacy logo has been removed.

schettn committed:
Improve documentation

The regions have been improved.

schettn committed:
Add attendee column to track list

Now the ohrwurm track list contains a new column
which lists all track attendees.

schettn committed:
Merge branch '003-add-orhwurm-track-attendees' into 004-improve-tracktable-styling

schettn committed:
Format track table created date column

Now the created date column is transformed to
a calendar format with momentJS.

schettn committed:
Improve track table attendee column styling

There is no wave animation on hover anymore.
Now there is a margin between single MDBChip
elements.

schettn committed:
Improve track table tag column styling

The shadow of the MDBBadge is removed now.
Also marings have been added.

schettn committed:
Improve track table transcript column styling

Now a MDBButton containing a MDBIcon is used instead
of a single MDBIcon.

schettn committed:
Merge pull request #5 from snek-shipyard/003-add-orhwurm-track-attendees

Add ohrwurm track attendees

schettn committed:
Merge pull request #6 from snek-shipyard/004-improve-tracktable-styling

Improve tracktable styling

pinterid committed:
Add track modal

The track modal was added.

schettn committed:
Change project and track add buttons to icons

Now a icon appears instead of a button with text.

pinterid committed:
Add track buttons

The track buttons for the list were added.

pinterid committed:
Add drag & drop for track list

Tracks can now be directly dragged into the list.

schettn committed:
Add password reset functionallity

Now it is possible to reset a user password.

pinterid committed:
Add project buttons

The project buttons for the list were added.

pinterid committed:
Merge remote-tracking branch 'origin/008-add-user-management' into 008-add-user-management

pinterid committed:
Add project modal

The project modal was added.

pinterid committed:
Add member management page

The member management page was added.

schettn committed:
Change project model user select style

Now the user selector and search bar are inside
a border.

pinterid committed:
Update button hitbox

The button hitbox was updated.

pinterid committed:
Update button hitbox

The button hitbox was updated.

pinterid committed:
Add modal styling

Styling for the modal was added.

pinterid committed:
Add member modal

The member modal was added.

pinterid committed:
Remove duplicated list

The track list appeared twice, this was fixed.

schettn committed:
Merge pull request #10 from snek-shipyard/007-add-track-upload

007 add track upload to next

pinterid committed:
Merge branch 'next' into 008-add-user-management

schettn committed:
Merge pull request #9 from snek-shipyard/008-add-user-management

008 add user management to next

pinterid committed:
Fix errors

Fixed some errors made during migration.

schettn committed:
Restructure components

Ohrwurm components have been restructered.

schettn committed:
Add project add/delete/update actions and expand pac query

Now actions for adding, deleting and updating tracks
are added. ChannelId and members are added to PAC.

schettn committed:
Add ohrwurm supervisor status to PAC member

Now the PAC members include a field to identify if
a member is a supervisor.

schettn committed:
Add fetch members action

Now it is possible to fetch members from the
backend.
Also the member type is added and integrated.

schettn committed:
Add member add action

Now it is possible to add members.

schettn committed:
Fix switch-case "or" violation

Because case requires constant expression as its value.
And since an || expression is not a compile time constant,
it is not allowed.
Ref: https://stackoverflow.com/questions/18412936/or-operator-in-switch-case

schettn committed:
Change id type and extend nesting

Now the ids are strings instead if number. 
Also some actions now return more information
of nested data.

schettn committed:
Add member delete action

The member delete functionality is now integrated
with the backend.

schettn committed:
Store generated password of member in store

Now after adding a new member the generated password
is stored.

schettn committed:
Add member update action

The member update functionality is now integrated
with the backend.

schettn committed:
Improve information storing on adding a new member

Now when a new member is added the username is also stored. Therefore the store structure has been improved.

schettn committed:
Fix id/index type mismatch

Now all indexes/ids are strings due to graphqls
ID type.

pinterid committed:
Add tag color picker

It is now possible to add tags with colors.

schettn committed:
Merge pull request #11 from snek-shipyard/008-add-backend-integration

008 add backend integration

pinterid committed:
Add track actions

The track actions, reducers and types were added.

schettn committed:
Fix id/index type mismatch

Now the indexes of the ProjectTable props functions are
of type string.

schettn committed:
Set maxHeight of tables full heigth

The maxHeight of member, project and track table
is set to 100% now.

pinterid committed:
Implement track add

It is now possible to add tracks to projects.

pinterid committed:
Implement track delete

It is now possible to delete tracks.

schettn committed:
Fix un-mutable track items array

Error: can't define array index property past the end of an array with non-writable length.

Now it is possible to add the added track to the list of
current tracks.

schettn committed:
Add supervisor status to authenticated user

Now the authenticated user contains information about
his supervisor status.

pinterid committed:
Implement track edit

It is now possible to edit tracks.

schettn committed:
Optimise track update

The values to be updated have been optimised.

pinterid committed:
Update track edit

The track edit was updated.

pinterid committed:
Merge remote-tracking branch 'origin/007-improve-track-upload' into 007-improve-track-upload

schettn committed:
Fix tags rendering issue

Now the tag color is always lowercased.

schettn committed:
Apply backend change for attendees

Now when creating and/or updating tracks
a list of attendee objects is now provided instead of
strings.

schettn committed:
Move transcript to track actions column

Now the transcript can be accessed in actions column.

pinterid committed:
Fix color tag

An issue concerning the color tag was fixed.

schettn committed:
Merge branch 'next' into 007-improve-track-upload

pinterid committed:
Fix migration

Some errors had to be fixed to merge.

schettn committed:
Merge pull request #12 from snek-shipyard/007-improve-track-upload

007 improve track upload to next

schettn committed:
Change old repo title

Steward is now changed to Watch.

pinterid committed:
Implement project add/delete

It is now possible to add and delete projects.

pinterid committed:
Implement project edit

It is now possible to edit projects.

schettn committed:
Prepare redux actions for member page

Now the member page includes all its required actions.

schettn committed:
Fix member page store connecting error

Now the member page is fully connected to redux.

schettn committed:
Update styling of all tables

Now there is a new styling for track, project and member
table.

schettn committed:
Add modify permissions for table actions

Now deleting and updating tracks/projects is only
possible if authenticated as supervisor.

pinterid committed:
Merge remote-tracking branch 'origin/next' into 008-integrate-member-project-actions-ui

schettn committed:
Enable track download option

Now it is possible to download the audio file of tracks.

schettn committed:
Improve wording

Now edit is used instead of view.

schettn committed:
Improve table action submenu colors

The colors of the table action have been changed.

pinterid committed:
Add member implementation

It is now possible to add, delete and update members.

schettn committed:
Add option to add/update pacs on member

Now it is possible to update the pacs of a member.

schettn committed:
Optimise navigation

The navbar component has been moved.
The navbar dropdown menu has been expanded.

schettn committed:
Improve routes

Now the members url is restricted to supervisor.

schettn committed:
Improve member table supervisor column

A check icon is rendered in the supervisor column if
a member is a supervisor.

schettn committed:
Improve members model

Now pacs are listed and searchable.
Adding them to a member is not supported yet.

schettn committed:
Fix member add/update bug

Now it is possible to add/update members again.

pinterid committed:
Add member project selector

It is now possible to select project in the member modal.

pinterid committed:
Add member search

It is now possible to search for members in the project modal.

pinterid committed:
Merge branch 'next' into 008-integrate-member-project-actions-ui

schettn committed:
Merge pull request #13 from snek-shipyard/008-integrate-member-project-actions-ui

008 integrate member project actions UI to next

schettn committed:
Improve wording

The view option on member page has been changed to edit.

schettn committed:
Code cleanup

The code taste has been improved.

schettn committed:
Add alert for generated password

Now when a new user is added the generated password is
alerted.

schettn committed:
Fix title/createdAt/description on track model

Now all fields of the track model are filled on edit.

schettn committed:
Revert "Publish docs build"

This reverts commit d6e7e2f8723bf1095eb6063b8afadd074637f4c2.

schettn committed:
Publish docs build

A test build is published.

schettn committed:
Add conditional rendering for navbar menu

Now the navbar menu is only visible if the authenticated
user is not anonymous.

pinterid commented on package.json:
```diff
@@ -3,11 +3,30 @@
     "react": "^16.13.1",
+    "react-data-table-component": "^6.11.6",
+    "react-dnd": "^5.0.0",
+    "react-dnd-html5-backend": "^3.0.2",
```
I thin these packages are no longer needed, because we use our own react-tag-input. Should be tested before merging.
```suggestion
```

pinterid commented on package.json:
```diff
@@ -3,11 +3,30 @@
-    "react-scripts": "^4.0.0"
+    "react-scripts": "^4.0.0",
+    "react-tag-input": "^6.5.2",
+    "reactjs-tag-input": "^2.0.15",
```
I thin these packages are no longer needed, because we use our own react-tag-input. Should be tested before merging.
```suggestion
```

pinterid commented on src/App.tsx:
```diff
@@ -18,24 +25,27 @@ import Routes from "./Routes";
-          <Footer />
-        </div>
-      </ScrollToTop>
+      <Navbar></Navbar>
```
```suggestion
      <Navbar />
```

pinterid commented on src/Routes.tsx:
```diff
@@ -3,28 +3,91 @@
+import MemberPage from "./components/pages/ohrwurm/MemberPage";
+
+import { RootState } from "./store/reducers/index";
+import { ThunkDispatch } from "redux-thunk";
```
Should be moved above the local imports.

pinterid commented on src/Routes.tsx:
```diff
@@ -3,28 +3,91 @@
+  Switch,
+  RouteComponentProps,
+  withRouter,
+} from "react-router-dom";
+//> Redux
+// Allows to React components read data from a Redux store, and dispatch actions
+// to the store to update data.
+import { connect } from "react-redux";
 
 //> Components
-import { HomePage } from "./components/pages";
+import { OhrwurmPage, OhrwurmMemberPage } from "./components/pages";
+import LoginPage from "./components/pages/LoginPage/index";
+import PasswordChangePage from "./components/pages/PasswordChangePage/index";
+import MemberPage from "./components/pages/ohrwurm/MemberPage";
+
+import { RootState } from "./store/reducers/index";
+import { ThunkDispatch } from "redux-thunk";
+import { UserState } from "./store/types";
+import { Ohrwurm } from "./components/organisms";
 //#endregion
 
+interface OwnProps {}
+interface StateProps {
+  user: UserState;
+}
+interface DispatchProps {}
+interface Props
+  extends OwnProps,
+    StateProps,
+    DispatchProps,
+    RouteComponentProps {}
+
```
```suggestion

#region > Interfaces
interface OwnProps {}
interface StateProps {
  user: UserState;
}
interface DispatchProps {}
interface Props
  extends OwnProps,
    StateProps,
    DispatchProps,
    RouteComponentProps {}
#endregion

```

pinterid commented on src/components/atoms/Breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,39 @@
+
+//#region > Exports
+export default Breadcrumbs;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/atoms/Mailto/index.tsx:
```diff
@@ -0,0 +1,31 @@
+
+//#region > Exports
+export default Mailto;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/atoms/index.ts:
```diff
@@ -1,12 +1,11 @@
-  ScrollToTop,
-};
+export { Breadcrumbs, Mailto, ScrollToTop };
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/Login/index.tsx:
```diff
@@ -0,0 +1,100 @@
+
+//#region > Exports
+export default LoginForm;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/PasswordChange/index.tsx:
```diff
@@ -0,0 +1,129 @@
+                      <p className="dark-grey-text font-weight-bold ml-1">
+                        {this.props.username}
+                      </p>
+                      .
```
Why is there a dot?

schettn replied:
Enter a new password for the user **pinterid** _._ <-- dot

pinterid commented on src/components/molecules/PasswordChange/index.tsx:
```diff
@@ -0,0 +1,129 @@
+    );
+  }
+}
+
+export default PasswordChangeForm;
```
```suggestion
#endregion

#region > Exports
export default PasswordChangeForm;
#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/ServiceGroup/index.tsx:
```diff
@@ -0,0 +1,68 @@
+
+//#region > Exports
+export default ServiceGroup;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/tables/MemberTable/index.tsx:
```diff
@@ -0,0 +1,115 @@
+
+//#region > Exports
+export default MemberTable;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/tables/ProjectTable/index.tsx:
```diff
@@ -0,0 +1,117 @@
+
+//#region > Exports
+export default ProjectTable;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/molecules/tables/TrackTable/index.tsx:
```diff
@@ -0,0 +1,200 @@
+
+//#region > Exports
+export default TrackTable;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/organisms/Navbar/index.tsx:
```diff
@@ -0,0 +1,132 @@
+import "./indes.scss";
+//> Logo
+import logo from "../../../assets/logo.png";
+
```
```suggestion
#endregion

```

pinterid commented on src/components/organisms/index.ts:
```diff
@@ -0,0 +1,10 @@
+
+//#region > Exports
+export { Navbar, Ohrwurm };
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/components/organisms/modals/TrackModal/index.tsx:
```diff
@@ -0,0 +1,416 @@
+  fetchPACSAction,
+  fetchPACTracksAction,
+} from "../../../../store/actions/ohrwurmActions";
+
```
```suggestion
```

pinterid commented on src/components/organisms/modals/TrackModal/index.tsx:
```diff
@@ -0,0 +1,416 @@
+    if (this.state.editing) {
+      const id = this.props.track?.id;
+
+      if (title === this.props.track?.title) {
+        title = undefined;
+      }
+      if (description === this.props.track?.description) {
+        description = undefined;
+      }
+      if (tags === this.props.track?.tags) {
+        tags = undefined;
+      }
```
```suggestion
      if (title === this.props.track?.title) {
        title = undefined;
      }
      
      if (description === this.props.track?.description) {
        description = undefined;
      }
      
      if (tags === this.props.track?.tags) {
        tags = undefined;
      }
```

pinterid commented on src/components/pages/ohrwurm/MemberPage/index.tsx:
```diff
@@ -0,0 +1,186 @@
+import { Breadcrumbs } from "../../../atoms";
+import { MemberTable } from "../../../molecules";
+import { MemberModal } from "../../../organisms/modals";
+
```
```suggestion
```

pinterid commented on src/global.d.ts:
```diff
@@ -0,0 +1 @@
+declare module "reactjs-tag-input";
```
Test the changes requested for the package.json, if the changes work this file should also be removed

pinterid commented on src/store/actions/ohrwurmActions.ts:
```diff
@@ -0,0 +1,1071 @@
+`;
+//#endregion
+
+//#region > Types
+type ohrwurmArguments = {
+  getClientSnek: () => SnekClient;
+};
+//#endregion
```
Move it to the types file?

pinterid commented on src/store/actions/ohrwurmActions.ts:
```diff
@@ -0,0 +1,1071 @@
+        payload: {
+          error: {
+            errorCode: 999,
+            message: `Updating Member failed (${ex.message})`,
```
```suggestion
            message: `Updating Track failed (${ex.message})`,
```

pinterid commented on src/store/actions/ohrwurmActions.ts:
```diff
@@ -0,0 +1,1071 @@
+        payload: {
+          error: {
+            errorCode: 999,
+            message: `Adding PAC failed (${ex.message})`,
```
```suggestion
            message: `Fetching Members failed (${ex.message})`,
```

pinterid commented on src/store/actions/userActions.ts:
```diff
@@ -0,0 +1,203 @@
+          message: "Logout failed",
+          error: ex,
+        },
+      });
```
```suggestion
      });
      
```

pinterid commented on src/store/actions/userActions.ts:
```diff
@@ -0,0 +1,203 @@
+          },
+          errorDetails: ex,
+        },
+      });
```
```suggestion
      });
      
```

pinterid commented on src/store/reducers/index.ts:
```diff
@@ -0,0 +1,34 @@
+//#endregion
+
+//#region > Exports
+export type RootState = ReturnType<typeof rootReducer>;
```
Move to types file?

schettn replied:
RootState should be part of reducers/index.ts

pinterid commented on src/store/reducers/index.ts:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * SPDX-License-Identifier: (EUPL-1.2)
+ * Copyright © 2019-2020 Simon Prast
```
```suggestion
 * Copyright © 2019-2020 Nico Schett
```

pinterid commented on src/store/reducers/userReducers.ts:
```diff
@@ -0,0 +1,75 @@
+
+//#region > Exports
+export default userReducer;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid commented on src/store/reducers/ohrwurmReducers.ts:
```diff
@@ -0,0 +1,101 @@
+
+//#region > Exports
+export default ohrwurmReducer;
+//#endregion
```
```suggestion
//#endregion

/**
 * SPDX-License-Identifier: (EUPL-1.2)
 * Copyright © 2019-2020 Nico Schett
 */
```

pinterid reviewed (changes requested):
Some changes should be made before merging.

schettn committed:
Apply suggestions from code review

Co-authored-by: Pinterics David <55298934+pinterid@users.noreply.github.com>

schettn committed:
Apply suggestions from code review

Co-authored-by: Pinterics David <55298934+pinterid@users.noreply.github.com>

schettn committed:
Update licenses

The date and owner of the copyright have been updated.

schettn committed:
Fix syntax error in new regions

The syntax errors in the new regions occurred through 
b577024b6b8ae92d2368c456fc65e97092e613e8 are fixed now.

pinterid committed:
Improve code quality

Some unneeded packages got removed.

schettn committed:
Improve routing and page restrictions/redirects

The routing and rendering process has been improved.

schettn committed:
Improve code quality

Some unused imports have been removed.

schettn committed:
Improve navbar conditional rendering of menu

Now the navbar menu is only rendered if the anonymous
status is false.

pinterid committed:
Update package.json

The package.json was updated to provide the deploy script.

pinterid reviewed (dismissed):
LGTM

pinterid reviewed (approved):
LGTM
